ecommerce/serializers.py

```diff
@@ -488,61 +489,88 @@ def update(self, instance, validated_data):
         if items is None and coupons is None and data_consents is None:
             raise ValidationError("Invalid request")
 
-        product, runs = self._update_items(basket, items)
-        coupon_version = self._update_coupons(basket, product, coupons)
-        with transaction.atomic():
-            if product:
-                # Update basket items and coupon selection
-                basket.basketitems.all().delete()
-                models.BasketItem.objects.create(
-                    product=product, quantity=1, basket=basket
-                )
+        if items == []:
+            self._update_basket_data(
+                basket, data_consents=data_consents, clear_all=True
+            )
+            return basket
 
-                models.CourseRunSelection.objects.filter(basket=basket).delete()
-                if runs is not None:
-                    for run in runs:
-                        models.CourseRunSelection.objects.create(basket=basket, run=run)
+        existing_product = basket.get_product()
+        coupon_version = None
 
-                if coupon_version:
-                    models.CouponSelection.objects.update_or_create(
-                        basket=basket, defaults={"coupon": coupon_version.coupon}
-                    )
-                else:
-                    basket.couponselection_set.all().delete()
-            else:
-                # Remove everything from basket
-                basket.basketitems.all().delete()
-                basket.couponselection_set.all().delete()
-                basket.courserunselection_set.all().delete()
+        if items is None:
+            updated_product = None
+            updated_run_ids = None
+            program_run = None
+        else:
+            updated_product, updated_run_ids, program_run = self._determine_items_to_update(
+                basket, items, existing_product
+            )
 
-            if data_consents is not None:
-                sign_date = datetime.now(tz=pytz.UTC)
-                models.DataConsentUser.objects.filter(
-                    id__in=data_consents, user=basket.user
-                ).update(consent_date=sign_date)
+        if updated_product or existing_product:
+            coupon_version = self._get_applicable_coupon_version(
+                basket,
+                # Whether or not the product was updated, we want to make sure that there is a valid
+                # coupon associated with the basket
+                product=updated_product or existing_product,
+                coupons=coupons,
+            )
+
+        self._update_basket_data(
+            basket,
+            updated_product=updated_product,
+            updated_program_run=program_run,
+            updated_run_ids=updated_run_ids,
+            coupon_version=coupon_version,
+            data_consents=data_consents,
+            clear_all=False,
+        )
 
         return instance
 
+    def _validate_runs(self, run_ids, product):
```

Comment:
I think this needs a `return {"runs": run_ids}` at the end?

Reply:
I initially tried making `validate_runs` a standard DRF validator function, but discovered that our front end depends on that validation being non-standard. Standard DRF validators return validation errors in a list. For example, in `validate_items`, `raise ValidationError("Invalid request")` results in a response body that looks like `{"errors": {"items": ["Invalid request"]}}`. Our code expects validation errors for `"runs"` to be a single string rather than a list. I didn't want to increase the scope of this PR too much, especially given the sorta-heavy refactor I already did on this serializer class, so I left this "non-standard" validator approach alone. If you see a simple way to get around that concern, let me know

Reply:
ok, that makes sense for what you're trying to do here